### PR TITLE
[GKD] Buffer Implementation for Distillation Trainer

### PR DESCRIPTION
**Pull request by cmpatino**
# Implement Buffer for Distillation Trainer (`GOLDTrainer`)

Implement generation buffering and multi-generation support for GOLDTrainer

Add a prompt-level generation buffer that decouples generation from the
optimization steps. We adopt a buffer similar to GRPO to generate all rollouts for all mini-batches within an optimization step, leveraging parallel inference engines. This means each worker handles a buffer of `per_device_train_batch_size * gradient_accumulation_steps`. 

## Buffer Details

We allow multiple rollouts per prompt, following Thinking Machine’s Tinker example. The number of rollouts per prompt is determined by the num_generations parameter. To keep the effective batch size constant, we introduce the generation_batch_size parameter, which controls how many unique prompts we pass to the inference engine. We enforce `generation_batch_size = per_device_train_batch_size * gradient_accumulation_steps // num_generations` to ensure the effective batch size is invariant across setups.

## Benchmarks

We can replicate Thinking Machine’s results using both non-Liger and Liger losses, achieving a 3x speedup on a setup with 8 training nodes in colocate mode.

comparison_tinker_liger

Phase | Tinker (s) | TRL (s)
-- | -- | --
Sampling | 329.83 | 130
Loss | 37.96 | -
Training | 98.69 | 38
Total | 492.28 | 173

## Before submitting
- [x] Did you read the [contributor guideline](https://github.com/huggingface/trl/blob/main/CONTRIBUTING.md#create-a-pull-request),
      Pull Request section?
- [x] Did you make sure to update the documentation with your changes?
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

---

> [!NOTE]
> **Medium Risk**
> Changes the GOLD training loop and dataloader sampling semantics to buffer generations across optimizer windows and support multiple rollouts per prompt, which can affect training dynamics and correctness across distributed setups.
> 
> **Overview**
> Adds an optimizer-window *generation buffer* to `GOLDTrainer`, decoupling on-policy rollout generation from individual gradient-accumulation microsteps. Training now samples a full window batch via a `RepeatSampler`, generates (optionally via vLLM) once per window, and reuses buffered slices across accumulation steps.
> 
> Extends `GOLDConfig` with `num_generations` and `generation_batch_size` (with strict partitioning validation) and adds `teacher_model_revision`; updates model/teacher revision handling in `gold.py` and teacher model instantiation. Generation/label masking is refactored to correctly handle left-padded prompts and completion-only outputs, and Liger+ZeRO-3 gets an explicit parameter gather context.
> 
> Updates docs to describe buffering behavior and config changes, and adds/adjusts tests to cover left-padding retokenization and prompt-length masking expectations.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit abfadc13ea8266503a6feb06a2b65c5fa453b5f9. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: trl/experimental/gold/gold.py [hunk 1]
```diff
@@ -78,6 +79,19 @@
     ################
     # Model & Tokenizer
     ################
+    if training_args.student_model_revision is None:
```
Comment (edbeeching) [inline trl/experimental/gold/gold.py]: Why do we need the `student_model_revision` parameter, can we not just use `model_revision` ?

File: trl/experimental/gold/gold_trainer.py [hunk 2]
```diff
@@ -223,6 +233,59 @@ def build_teacher_inputs_from_texts(
     return teacher_input_ids, teacher_labels, teacher_attention_mask, teacher_prompt_length
 
 
+class _RepeatEachBatchDataLoader:
```
Comment (edbeeching) [inline trl/experimental/gold/gold_trainer.py]: Is this the same as the GRPO trainer one? can they be shared?

**Comment by lewtun** [comment]
@codex review

File: trl/experimental/gold/gold_trainer.py (hunk 2 above)
Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py]: The main reason to use `_RepeatEachBatchDataLoader` is that it doesn't re-collate samples that will be effectively ignored during training, and it handles dropping partial batches.

Both `RepeatSampler` (the one used in GRPO) and `_RepeatEachBatchDataLoader` repeat the batch to get the right number of optimization steps from the `transformers` Trainer. These repeated batches are ignored because the samples are taken from the buffer instead of from the repeated batch.

Looking back, I think the efficiency gains and handling the partial batches are negligible compared to keeping the codebase tidy without extra classes.

Do you agree with dropping the extra class and using `RepeatSampler` instead?

File: trl/experimental/gold/gold.py (hunk 1 above)
Comment (cmpatino) [reply trl/experimental/gold/gold.py]: True. I'll support just `model_revision`

File: trl/experimental/gold/gold_trainer.py (hunk 2 above)
Comment (edbeeching) [reply trl/experimental/gold/gold_trainer.py]: If you can get comparable results / speed then I agree to keep the codebase leaner

File: trl/experimental/gold/gold_trainer.py [hunk 3]
```diff
@@ -1044,6 +1064,504 @@ def _set_signature_columns_if_needed(self):
...
+                temperature,
+                top_k,
+                top_p,
+                repetition_penalty,
+                min_p,
+                n=self.num_generations,
+            )
+        else:
+            self._generate_non_vllm_for_slices(slices, on_policy_indices)
+            return
```
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py:1290]: This is not actually a bug. The block inside the else statement is only reached when `use_vllm=False`.

File: trl/experimental/gold/gold_trainer.py [hunk 4]
```diff
@@ -1044,6 +1116,508 @@ def _set_signature_columns_if_needed(self):
...
+                    completion_ids_for_text.append(completion_tensor.tolist())
+                else:
+                    padded_completion_ids_list.append(completion_tensor)
+                    completion_ids_for_text.append(completion_tensor.tolist())
+
+            completion_ids_padded = torch.stack(padded_completion_ids_list)
+
+            new_input_ids = torch.cat([prompt_ids, completion_ids_padded], dim=1)
+            prompt_lengths = (prompt_ids != pad_token_id).sum(dim=1)
+            new_attention_mask, new_labels = self._build_sequence_batch(
```
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py]: Fixed in [c3a8d73](https://github.com/huggingface/trl/pull/5137/commits/c3a8d73f45a28f347bac8dcc17ec4b35dc85de2d)

File: trl/experimental/gold/gold_trainer.py [hunk 5]
```diff
@@ -1634,17 +2188,7 @@ def generate_on_policy_outputs(self, model, inputs, generation_config, pad_token
...
-            new_attention_mask[new_input_ids == pad_token_id] = 0
-
-        new_labels = torch.full_like(new_input_ids, -100)
-        for idx in range(batch_size):
-            length = int(prompt_lengths[idx].item())
-            new_labels[idx, length:] = new_input_ids[idx, length:]
-
-        if pad_token_id is not None:
-            new_labels[new_input_ids == pad_token_id] = -100
+        new_attention_mask, new_labels = self._build_sequence_batch(new_input_ids, prompt_lengths, pad_token_id)
```
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py]: Fixed in [c3a8d73](https://github.com/huggingface/trl/pull/5137/commits/c3a8d73f45a28f347bac8dcc17ec4b35dc85de2d)

File: trl/experimental/gold/gold.py [hunk 6]
```diff
@@ -93,21 +93,21 @@
     if training_args.teacher_tokenizer_name_or_path is None and training_args.use_uld_loss:
         training_args.teacher_tokenizer_name_or_path = training_args.teacher_model_name_or_path
     teacher_model_kwargs = dict(
-        revision=model_args.model_revision,
```
[the remark this answers is not in the document]
Comment (cmpatino) [reply trl/experimental/gold/gold.py:96]: Addressed in [30a0fd5](https://github.com/huggingface/trl/pull/5137/commits/30a0fd564d9b341061491b64aa32608df2f4e9eb).

I added a new arg that allows the user to specify a different revision for the teacher. If it doesn't specify one, it will be the `main` revision.

File: trl/experimental/gold/gold_trainer.py [hunk 7]
```diff
@@ -1044,6 +1063,505 @@ def _set_signature_columns_if_needed(self):
...
+                        gen_prompts, dedup_mapping = dedup_result
+                        gen_n = n
+                    else:
+                        gen_prompts = all_prompts_text
+                        gen_n = 1
+                else:
+                    gen_prompts = all_prompts_text
+                    gen_n = 1
+
+                completion_ids = self.vllm_client.generate(
```
Comment (qgallouedec) [inline trl/experimental/gold/gold_trainer.py:1352]: ideally we would like to use `trl.generation.VLLMGeneration`, but it can be done in a future PR if it's too hard

**Comment by qgallouedec** [comment]
I haven't reviewed it in detail; I have a general idea of what it's about, but I'm leaving the implementation mostly up to you. In future PRs, we can try to align it better with the rest of the codebase, but what matters most right now are the results you're getting.

Make sure to run `make precommit` to make to CI happy

File: docs/source/gold_trainer.md [hunk 8]
```diff
@@ -29,6 +29,10 @@ messages). Important configuration flags on [`GOLDConfig`] include:
   matched/unmatched loss.
 * `beta`, `lmbda`, `seq_kd` – inherited from [`experimental.gkd.GKDConfig`], controlling the generalized JSD interpolation and on-policy
   sampling ratio.
+* `num_generations`, `generation_batch_size` – control buffered rollout generation across gradient accumulation windows.
+  `generation_batch_size` is the number of unique prompts per worker per optimizer step.
+* `student_model_revision` and `model_revision` – if `student_model_revision` is unset, GOLD uses `model_revision`.
```
Comment (qgallouedec) [inline docs/source/gold_trainer.md]: student_model_revision is removed in this pr no?

File: trl/experimental/gold/gold_config.py [hunk 9]
```diff
@@ -365,12 +382,6 @@ class GOLDConfig(SFTConfig):
     num_completions_to_print: int = field(default=5, metadata={"help": "Number of completions to print."})
```
Comment (qgallouedec) [inline trl/experimental/gold/gold_config.py]:
```suggestion
```

duplicated

**Review by qgallouedec — APPROVED** [review]
just ensure the CI is green before merging

File: trl/experimental/gold/gold_trainer.py (hunk 7 above)
Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py:1352]: I'll make sure to upload a PR with this change.

Comment (cmpatino) [reply trl/experimental/gold/gold_trainer.py:1352]: Addressed in https://github.com/huggingface/trl/pull/5351
